src/msssim.jl

```diff
@@ -84,8 +83,13 @@ function _msssim_map(iqi::MSSSIM, x::GenericGrayImage, ref::GenericGrayImage)
         imfilter(ref, window, "symmetric")
 
         # downsampling
-        x = x[1:2:end, 1:2:end]
-        ref = ref[1:2:end, 1:2:end]
+        if (ndims(x)==2)
+            x = x[1:2:end, 1:2:end]
+            ref = ref[1:2:end, 1:2:end]
+        else
+            x = x[1:end,1:2:end, 1:2:end]
+            y = ref = ref[1:end, 1:2:end, 1:2:end]
+        end
```

Comment:
IIRC SSIM is not limited to 2D matrix, right? This manually reads a 3D gray image as CHW and thus makes the code not so generic.

If RGB images are processed channel by channel, probably it would be better to split it before passing into the gray implementation, e.g., it could be:

```julia
function _msssim_map(iqi::MSSSIM, x::AbstractArray{<:AbstractRGB}, ref::AbstractArray{<:AbstractRGB})
    map(eachslice(1, channelview(x)), eachslice(1, channelview(ref))) do cx, cy
        _mssim_map(iqi, cx, cy)
    end
end
```

Reply:
Two things I'm not sure of here:
1. I tried separating the calls for RGB into separate calls for each of the channels and then taking the mean of the 3 results I got, but the answers differ from the current implementation (by 0.004 on one of the testcases, not sure of how significant that it). This is because `_ssim_map` itself accepts RGB images by just breaking it into a 3D array:
```(julia)
_ssim_map(iqi::SSIM, x::AbstractArray{<:AbstractRGB}, ref::AbstractArray{<:AbstractRGB}, peakval = 1.0, K = SSIM_K) =
    _ssim_map(iqi, channelview(x), channelview(ref), peakval, K)
```

2. Even if we do fix 1. and handle RGB images, I believe that `GenericGrayImage` allows the image to be >2 dimensional (something like a 3D gray image), in which case we'll still have to worry about not changing the "channel" dimension while downsampling, what the snippet above was doing

@johnnychen94

Reply:
I see, SSIM is done this way to match the results of other SSIM implementations. `0.004` should be considered a large difference; it's usually reported to precision `0.001`.

Sounds like some code replicate is unavoidable in RGB method to make MSSSIM match the results of other implementations.

Reply:
Not so sure how to proceed here :sweat_smile:. Should we let the `_msssim_map` function handle >2 dimensions (whether it is a 3D gray image, or the different channels of RGB) and the RGB method just breaks down the RGB pixels into separate dimensions, as was before?

Reply:
https://github.com/JuliaImages/ImageQualityIndexes.jl/blob/bb12ea3b8f2181161947d4f3b5ba3972ff2a5e19/src/msssim.jl#L43-L103 Copying and mosifying a RGB-specific version of it to `_msssim_map(iqi::MSSSIM, x::AbstractArray{<:AbstractRGB}, ref::AbstractArray{<:AbstractRGB})` would always work. 

Yes, there're some other details that need to be handled, e.g., minimizing the code redundancy, check if 3D RGB image works as expected.

Reply:
Because `SSIM` expects RGB images broken down into `CHW` views, and returns one value for all three of those channels, we can't process the channels separately. Hence, currently `_msssim_map(iqi::MSSSIM, x::AbstractArray{<:AbstractRGB}, ref::AbstractArray{AbstractRGB})` simply breaks down the RGB pixels into 3 channels of gray and passes onto `_msssim_map`.

A 3D gray image is already broken down this way, and hence works fine with `_msssim_map`.
The `if` condition here just takes care of the dimension while downsampling:
```
        if (ndims(x)==2)
            x = x[1:2:end, 1:2:end]
            ref = ref[1:2:end, 1:2:end]
        else
            x = x[1:end,1:2:end, 1:2:end]
            y = ref = ref[1:end, 1:2:end, 1:2:end]
        end
```

I am not sure of how this should be handled differently, or what generic case does it not consider. What am I missing here? :sweat_smile: 
@johnnychen94

Reply:
> A 3D gray image is already broken down this way, and hence works fine with _msssim_map.

I haven't yet read the details of MSSSIM so I could be wrong about this. But I think the general version for gray image should be written as:

```julia
subindices = ntuple(i->1:2:size(x, i), ndims(x))
x = x[subindices...]
ref = ref[subindices...]
```

Except for the RGB compatibility, I don't understand why the first dimension is handled specially.

---

I'm a little bit too busy these days so I didn't read the details, it's very possible that I've made mistakes here.

Reply:
Finally, that made sense to me!

What I just pushed tries to make `_msssim_map` agnostic to the colorant information. This is because `_msssim_map` does not perform any operations on the pixels directly, but simply delegates that to `SSIM` and `imfilter`, both of which know how to handle `RGB`, `Gray` or any other type of images. `_msssim_map` only cares about the positions of pixels in the array (for example while downsampling), and hence letting one pixel denote one location in the image always makes sense.

Correct me if I'm wrong on this, but because `channelview` (used in `_ssim_map` under the hood for colored images) does not copy data but only reinterprets a block of memory, repeated calls to it (per level of msssim) shouldn't cause performance issues.
@johnnychen94